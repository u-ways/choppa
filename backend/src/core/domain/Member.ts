export default interface Member {
  readonly id: string;
  readonly name: string;
  // readonly role: Chapter;
}
