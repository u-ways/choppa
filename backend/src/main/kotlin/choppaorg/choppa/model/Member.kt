package choppaorg.choppa.model

import javax.persistence.*

@Entity
class Member(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Int?,

    var name: String,

    @OneToOne
    @JoinColumn(name = "chapter_id")
    var chapter: Chapter?
) {
    constructor(name: String, chapter: Chapter): this(null, name, chapter)
}