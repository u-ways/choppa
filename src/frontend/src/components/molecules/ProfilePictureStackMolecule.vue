<template>
  <div class="profile-picture-stack row m-0">
    <ProfilePictureAtom v-for="member in membersShowing"
                        v-bind:key="member.id"
                        :member="member"
                        cssClasses="profile-picture-stack__picture"
    />
    <div v-if="numberOfExtraMembers > 0" class="profile-picture-stack__not-showing-count">
      +{{ numberOfExtraMembers }}
    </div>
  </div>
</template>

<script>
import ProfilePictureAtom from "@/components/atoms/ProfilePictureAtom";

export default {
  name: "ProfilePictureStackMolecule",
  components: { ProfilePictureAtom },
  props: {
    members: Array,
    showCount: {
      type: Number,
      default: 3,
    },
  },
  computed: {
    membersShowing() {
      return this.members.slice(0, this.showCount);
    },
    numberOfExtraMembers() {
      return this.members.length - this.showCount;
    },
  },
};
</script>

<style scoped lang="scss">
@import "../../assets/scss/colours";
@import "../../assets/scss/typography";

.profile-picture-stack {

  & > div {
    margin-left: -0.55rem;
  }

  & > div:first-of-type {
    margin-left: 0;
  }

  &__column {
    display: flex;
  }

  &__picture {
    width: 1.5rem;
    height: 1.5rem;
    display: inline-block;
  }

  &__not-showing-count {
    width: 1.5rem;
    height: 1.5rem;
    line-height: 1.5rem;
    border-radius: 50%;
    background: $white;
    border: 1px solid $white-border;
    text-align: center;
    vertical-align: middle;
    color: $text-dark;
    font-weight: $font-weight-light;
    font-size: 0.75rem;
    z-index: 1;
  }
}
</style>
