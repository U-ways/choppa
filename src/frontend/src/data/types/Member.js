export default class Member {
  constructor(id, name, chapter) {
    this._id = id;
    this._name = name;
    this._chapter = chapter;
  }

  get id() {
    return this._id;
  }

  get name() {
    return this._name;
  }

  set name(newName) {
    this._name = newName;
  }

  get chapter() {
    return this._chapter;
  }
}
