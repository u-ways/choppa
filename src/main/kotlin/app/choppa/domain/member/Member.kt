package app.choppa.domain.member

import app.choppa.domain.base.BaseModel
import app.choppa.domain.chapter.Chapter
import app.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import app.choppa.domain.squad.Squad
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @Column(name = "active", columnDefinition = "BOOLEAN")
    val active: Boolean = true,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=${chapter.name})"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
