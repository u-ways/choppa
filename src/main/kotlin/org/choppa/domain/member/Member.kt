package org.choppa.domain.member

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.chapter.Chapter
import org.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "member")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
