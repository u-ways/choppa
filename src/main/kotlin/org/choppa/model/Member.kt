package org.choppa.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.relations.IterationHistory
import org.choppa.model.relations.SquadCurrentMembers
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @OneToMany(mappedBy = "member", fetch = EAGER)
    @JsonIgnore
    var squads: MutableList<SquadCurrentMembers> = mutableListOf(),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    var iterations: MutableList<IterationHistory> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
