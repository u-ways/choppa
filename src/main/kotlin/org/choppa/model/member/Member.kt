package org.choppa.model.member

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.chapter.Chapter
import org.choppa.model.history.History
import org.choppa.model.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @ManyToMany(mappedBy = "members", fetch = EAGER)
    @JsonIgnore
    val squads: MutableList<Squad> = mutableListOf(UNASSIGNED_SQUAD),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    val iterations: MutableList<History> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
