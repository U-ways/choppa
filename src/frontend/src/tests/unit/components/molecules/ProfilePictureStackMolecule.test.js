import { describe, expect, it, shallowMount } from "@/tests/unit/test-imports";
import ProfilePictureStackMolecule from "@/components/molecules/ProfilePictureStackMolecule";
import Member from "@/data/types/Member";

const memberJohnny = new Member("1", "Johnny");
const memberRobert = new Member("2", "Robert");
const memberLucie = new Member("3", "Lucie");
const memberPeter = new Member("4", "Peter");
const memberKelly = new Member("5", "Kelly");
const members = [memberJohnny, memberRobert, memberLucie, memberPeter, memberKelly];

describe("Profile Picture Stack Molecule test", () => {
  it("Shows the correct amount of members", () => {
    expect(ProfilePictureStackMolecule.computed.membersShowing.call({
      members,
      showCount: 2,
    })).toIncludeAllMembers([memberJohnny, memberRobert]);

    expect(ProfilePictureStackMolecule.computed.membersShowing.call({
      members,
      showCount: 3,
    })).toIncludeAllMembers([memberJohnny, memberRobert, memberLucie]);

    expect(ProfilePictureStackMolecule.computed.membersShowing.call({
      members,
      showCount: 0,
    })).toBeArrayOfSize(0);
  });

  it("Renders correctly", async () => {
    const wrapper = shallowMount(ProfilePictureStackMolecule, {
      propsData: {
        members,
        showCount: 3,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
