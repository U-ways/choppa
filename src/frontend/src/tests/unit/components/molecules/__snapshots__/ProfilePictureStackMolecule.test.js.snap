// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Picture Stack Molecule test Renders correctly 1`] = `
"<div class=\\"profile-picture-stack row m-0\\">
  <profilepictureatom-stub member=\\"[object Object]\\" css=\\"profile-picture-stack__picture\\"></profilepictureatom-stub>
  <profilepictureatom-stub member=\\"[object Object]\\" css=\\"profile-picture-stack__picture\\"></profilepictureatom-stub>
  <profilepictureatom-stub member=\\"[object Object]\\" css=\\"profile-picture-stack__picture\\"></profilepictureatom-stub>
  <div class=\\"profile-picture-stack__not-showing-count\\">
    +2
  </div>
</div>"
`;
